dev/camdenhells/REQUIREMENTS.md

```diff
@@ -324,41 +350,75 @@ This does not preclude performance impacts; clearly there will be limits, but ha
-3. Compile-time API-usage validation, e.g. annotations ?
-  * Are these enforceable e.g. via lint?
-4. Runtime API-usage validation for the app, e.g. debug-only assertions/exceptions?
-  * 
-5. Debug visualizations (#1436)
-  * Trident, camera eyeball/frustum, lighting rays/lines, object bounds
-  * Are there (non-debug) app use cases for some of these as well? e.g.: 
-    * Interactive camera/lighting selections from a global-view PIP inset 
-    * Theater/stage lighting design apps
-    * Interior design apps - door/window/furniture/light placements
-  * If so perhaps just call them e.g. camera/light/bounds visualizations  
+  * Provide a variant that includes queueing to main/UI thread (see above)
+
+# 4. Testing, Validation, and Debugging Requirements
+
+## 4.1 Compile-time
+
+  * Consistently specify any constraints on all member and parameter values
+    * Specify all references as either nullable or non-null
+    * Specify integer and float ranges whenever less than built-in language limits
+  * ~Validate public API-usage [2.1]~
+     * ~Create an Android lint plugin to detect invocations of non API methods?~
+     * Seems like a reliable library-specific tool would be a lot of work/maintenance, and unlikely that client developers would download/install/use it anyway
+
+## 4.2 Build-time 
+
+  Help! I'm trying to define the line on what should be unit tested and what should not, and don't even understand the hardware/software environment in which unit tests run (i.e. emulators use what version of OpenGL? is there an Android process/activity context? etc)
+  
+  * Automated unit/integration test harness and tests 
+  * For all public API and internal interface methods except those:
+    * With significant historical runtime event dependencies or behavioral states, and/or
+    * Whose success cannot be (at least partially) reasonably determined as value assertions
+  * Includes e.g. contructors, simple get/set accessors, deterministic functions/transforms
+  * Excludes e.g. frame draws, client callbacks
+  
+## 4.3. Run-time
+
+For engine and app development and integration testing, and for debugging support issues
+
+1. Provide unified logging
+  * The usual levels - verbose, debug, info, warning, error, wtf
+  * All messages should identify the class and method
+  * Core engine/rendering component (scenes, scene view, render pass, etc.) messages should also identify the instance
+  * All warning and error messages should include specific causes and data values if appropriate
+  * All error messages should be accompanied by a thrown exception
+2. Provide debug-level method tracing messages 
+  * For key lifecycle event (create/init/start/run/pause/stop/destroy etc.) handlers of engine/rendering components
+    * Includes all active components: scenes, scene views, render passes, etc.
+    * Does not include passive model objects, textures, resources 
+  * Tracing dynamically enabled per instance; disabled by default  
```

Comment:
Can you clarify what you mean by this?

Reply:
I mean there would be e.g. a simple boolean for each core control/rendering component that determines whether any trace messages are actually logged for that instance; the trace methods for an instance would all be conditional. So you could just enable tracing on this scene, or that scene view, or that render pass. Most of the time there would be no trace output.